jaeger-thrift/src/main/java/io/jaegertracing/thrift/internal/senders/HttpSender.java

```diff
@@ -119,13 +134,98 @@ public Builder withAuth(String authToken) {
+      pins.addAll(Arrays.asList(sha256certs));
+      return this;
+    }
+
+    /**
+     * Enable accepting self-signed certificates. This will only take effect if pinning is provided.
+     */
+    public Builder acceptSelfSigned() {
+      this.selfSigned = true;
+      return this;
+    }
+
     public HttpSender build() {
       if (authInterceptor != null) {
         clientBuilder.addInterceptor(authInterceptor);
       }
+      try {
+        // Just obtain hostname for SSL feature. Failure is OK if plain http is used.
+        final URI uri = new URI(endpoint);
+        String hostname = hostname = uri.getHost();
+
+        if (!selfSigned && !pins.isEmpty()) {
+          // Pinning Certificate issued by public CA
+          for (String cert: pins) {
+            certificatePinnerBuilder.add(hostname, String.format("%s", cert));
+          }
+          clientBuilder.certificatePinner(certificatePinnerBuilder.build());
+        } else if (selfSigned && !pins.isEmpty()) {
+          /* Issued by private CA, OkHttp's pinner is unable to verify the pins.
+           * Instead, check the sha256 hash value by custom verifier. */
+          acceptSelfSigned(clientBuilder, hostname, pins);
+        }
+      } catch (Exception e) {
+        // Won't validate the endpoint name.
+        // True moment comes at Sender's Build.
+      }
       return new HttpSender(this);
     }
 
+    private void acceptSelfSigned(OkHttpClient.Builder clientBuilder, final String hostname, final List<String> pinlist) {
+      try {
+        final TrustManager[] selfSignedServerTrustManager = new TrustManager[] {
+          new X509TrustManager() {
+            private final String subjectCN = "CN=" + hostname;
+            private final List<String> pins = pinlist;
+
+            private boolean check(X509Certificate[] chain) {
+              // Intersection of pins and every cert in this chain
+              for (X509Certificate cert: chain) {
+                String hash = CertificatePinner.pin(cert);
+                for (String pin: pins) {
+                  if (hash.equals(pin)) {
+                    return true;
+                  }
+                }
+              }
+              return false;
+            }
+
+            @Override public void checkServerTrusted(X509Certificate[] chain, String authType) throws CertificateException {
+              for (X509Certificate cert: chain) {
+                String[] subject = cert.getSubjectDN().getName().split("/");
+                for (String name: subject) {
+                  if (subjectCN.equals(name)) {
+                    // Found endpoint's hostname. This chain is going to be tested.
+                    if (check(chain)) {
+                      return;
+                    }
+                  }
+                }
+              }
+              // TODO: For TOFU (trust on first use) usecase, print the chain
+              throw new CertificateException();
```

Comment:
If there's anything wrong, it would be very hard to figure out *what* is wrong. I see the following possible error conditions:

* Empty chain (bad configuration from the user's part)
* No matches between the hostname and the subject CNs from the chain
* The check failed

It would be good to have an exception thrown with the appropriate message, even if it's a bit more code.

Reply:
`CertificateException` is thrown when the certificate couldn't verified _by the `TrustManager`_. If another (like default) manager might succeed the verification even after _this_ manager failed, the connection must be sane.

To follow this (strange) interface, I throw `CertificateException` whenever this manager fails to verify the connection.

But probably other exception could be thrown to fail in case of "The check failed" but CN matches including server's chain was empty.

Reply:
> But probably other exception could be thrown to fail in case of "The check failed" but CN matches including server's chain was empty.

This is good for trust-on-first-use use-case to print actual pins of the server, because calculating the pin by `openssl` command is a little bothering.

Reply:
+1 for having different exceptions for the `check failed` case